src/ProducerEventProcessor/UploaderCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/Logging.h"
+#include "ProducerEventProcessor/UploaderCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+namespace {
+
+static constexpr const char* kAnswerString = "Test Interned String";
+static constexpr uint64_t kAnswerKey = 42;
+
+static ClientCaptureEvent CreateInternedStringCaptureEvent(uint64_t key, const std::string& str) {
+  ClientCaptureEvent event;
+  event.mutable_interned_string()->set_key(key);
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+static ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderCaptureEventCollectorTest : public testing::Test {
+ public:
+  void Start(uint64_t event_count, uint64_t produce_duration_ms, uint64_t upload_duration_ms) {
+    event_count_ = event_count;
+    produce_duration_ms_ = produce_duration_ms;
+    upload_duration_ms_ = upload_duration_ms;
+
+    wait_all_data_uploaded_duration_ms_ =
+        upload_duration_ms_ * 2 + produce_duration_ms_ * event_count_;
+
+    data_produce_thread_ =
+        std::thread(&UploaderCaptureEventCollectorTest::ProduceCaptureData, this);
+    data_upload_thread_ = std::thread(&UploaderCaptureEventCollectorTest::UploadCaptureData, this);
+  }
+
+  void Stop() {
+    absl::MutexLock lock{&mutex_};
+    stop_called_ = true;
+  }
+
+ protected:
+  void TearDown() override {
+    absl::MutexLock lock{&mutex_};
+    CHECK(stop_called_);
+
+    if (data_produce_thread_.joinable()) data_produce_thread_.join();
+    if (data_upload_thread_.joinable()) data_upload_thread_.join();
+  }
+
+  UploaderCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+  uint64_t wait_all_data_uploaded_duration_ms_;
+
+ private:
+  // Keep producing capture event until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the capture finish event.
+  void ProduceCaptureData() {
+    size_t event_count = event_count_;
+    bool stopped = false;
+    while (event_count-- > 1 && !stopped) {
+      mutex_.LockWhenWithTimeout(absl::Condition(&stop_called_),
+                                 absl::Milliseconds(produce_duration_ms_));
```

Comment:
Ok, so an event is produced every `produce_duration_ms_`? From the name I thought this meant that the total duration of this thread was `produce_duration_ms_`. I think this field should be called `produce_event_every_ms_` instead (or something like that).

Reply:
Make sense. Renamed now. Thanks!